File: Sources/PalmierPro/Editor/RippleEngine.swift
import Foundation

/// Pure functions for ripple editing: computing how clips shift after
/// insertions or deletions. No state, easily testable.
enum RippleEngine {

    /// After removing clips from a track, compute new start frames for
    /// remaining clips that should shift backward to close the gap.
    ///
    /// Returns [(clipId, newStartFrame)] for clips that need to move.
    static func computeRippleShifts(
        clips: [Clip],
        removedIds: Set<String>
    ) -> [(clipId: String, newStartFrame: Int)] {
        let removedRanges = clips
            .filter { removedIds.contains($0.id) }
            .map { (start: $0.startFrame, end: $0.endFrame) }
        return computeRippleShiftsForRanges(
            clips: clips.filter { !removedIds.contains($0.id) },
            removedRanges: removedRanges
        )
    }

    /// Shift clips leftward to close the gaps defined by `removedRanges`.
    /// Used when ranges come from a different track (sync-locked ripple).
    static func computeRippleShiftsForRanges(
        clips: [Clip],
        removedRanges: [(start: Int, end: Int)]
    ) -> [(clipId: String, newStartFrame: Int)] {
        let merged = mergeRanges(removedRanges)
        guard !merged.isEmpty else { return [] }

        var shifts: [(clipId: String, newStartFrame: Int)] = []
        for clip in clips.sorted(by: { $0.startFrame < $1.startFrame }) {
            var shift = 0
            for range in merged where range.end <= clip.startFrame {
                shift += range.end - range.start
            }
            if shift > 0 {
                shifts.append((clip.id, clip.startFrame - shift))
            }
        }
        return shifts
    }

    /// Push all clips at or after `insertFrame` forward by `pushAmount` frames.
    /// Returns [(clipId, newStartFrame)] for clips that need to move.
    static func computeRipplePush(
        clips: [Clip],
        insertFrame: Int,
        pushAmount: Int,
        excludeIds: Set<String> = []
    ) -> [(clipId: String, newStartFrame: Int)] {
        var shifts: [(clipId: String, newStartFrame: Int)] = []
        for clip in clips where !excludeIds.contains(clip.id) {
            if clip.startFrame >= insertFrame {
                shifts.append((clip.id, clip.startFrame + pushAmount))
            }
        }
        return shifts
    }

    // MARK: - Helpers

    private static func mergeRanges(_ ranges: [(start: Int, end: Int)]) -> [(start: Int, end: Int)] {
        let sorted = ranges.sorted { $0.start < $1.start }
        var merged: [(start: Int, end: Int)] = []
        for range in sorted {
            if let last = merged.last, range.start <= last.end {
                merged[merged.count - 1] = (last.start, max(last.end, range.end))
            } else {
                merged.append(range)
            }
        }
        return merged
    }
}

File: platforms/windows/src/PalmierPro.Core/Timeline/RippleEngine.cs
using PalmierPro.Core.Models;

namespace PalmierPro.Core.Timeline;

/// A proposed new start frame for a single clip, produced by the ripple engine and applied by the
/// caller. Ported from Editor/RippleEngine.swift.
public readonly record struct ClipShift(string ClipId, int NewStartFrame);

/// A half-open `[Start, End)` frame interval on a single track.
public readonly record struct FrameRange(int Start, int End)
{
    public int Length => End - Start;
}

/// A user-selected empty gap on a single track.
public readonly record struct GapSelection(int TrackIndex, FrameRange Range);

/// Pure functions for ripple editing: computing how clips shift after insertions or deletions.
/// Ported near-verbatim from Editor/RippleEngine.swift.
public static class RippleEngine
{
    /// After removing clips from a track, compute new start frames for remaining clips that
    /// should shift backward to close the gap.
    public static List<ClipShift> ComputeRippleShifts(IReadOnlyList<Clip> clips, ISet<string> removedIds)
    {
        var removedRanges = clips
            .Where(c => removedIds.Contains(c.Id))
            .Select(c => new FrameRange(c.StartFrame, c.EndFrame))
            .ToList();
        return ComputeRippleShiftsForRanges(clips.Where(c => !removedIds.Contains(c.Id)).ToList(), removedRanges);
    }

    /// Shift clips leftward to close the gaps defined by `removedRanges`. Used when ranges come
    /// from a different track (sync-locked ripple).
    public static List<ClipShift> ComputeRippleShiftsForRanges(IReadOnlyList<Clip> clips, IReadOnlyList<FrameRange> removedRanges)
    {
        var merged = MergeRanges(removedRanges);
        if (merged.Count == 0)
        {
            return [];
        }

        var shifts = new List<ClipShift>();
        foreach (var clip in clips.OrderBy(c => c.StartFrame))
        {
            var shift = merged.Where(r => r.End <= clip.StartFrame).Sum(r => r.Length);
            if (shift > 0)
            {
                shifts.Add(new ClipShift(clip.Id, clip.StartFrame - shift));
            }
        }
        return shifts;
    }

    /// Push all clips at or after `insertFrame` forward by `pushAmount` frames.
    public static List<ClipShift> ComputeRipplePush(
        IReadOnlyList<Clip> clips, int insertFrame, int pushAmount, ISet<string>? excludeIds = null)
    {
        excludeIds ??= new HashSet<string>();
        return clips
            .Where(c => !excludeIds.Contains(c.Id) && c.StartFrame >= insertFrame)
            .Select(c => new ClipShift(c.Id, c.StartFrame + pushAmount))
            .ToList();
    }

    public static List<FrameRange> MergeRanges(IReadOnlyList<FrameRange> ranges)
    {
        var sorted = ranges.OrderBy(r => r.Start).ToList();
        var merged = new List<FrameRange>();
        foreach (var range in sorted)
        {
            if (merged.Count > 0 && range.Start <= merged[^1].End)
            {
                merged[^1] = new FrameRange(merged[^1].Start, Math.Max(merged[^1].End, range.End));
            }
            else
            {
                merged.Add(range);
            }
        }
        return merged;
    }
}
